### PR TITLE
CDAP-15719 Multi Table source to support partitioning

### PR DESCRIPTION
**Pull request by mlozbin-cybervisiontech**
JIRA: https://issues.cask.co/browse/CDAP-15719
WIKI: https://wiki.cask.co/display/CE/MultiTableDatabase+Batch+Source
In scope of this PR:

- Added table inner splitting like in Database plugins
- Fixed tests

### CODE REVIEW

**Comment by mlozbin-cybervisiontech** [comment]
@albertshau could you, please, take a look.

File: docs/MultiTableDatabase-batchsource.md [hunk 1]
```diff
@@ -51,6 +51,7 @@ By default reads all tables matching the tableNamePattern.
 **blackList**: Used in conjunction with tableNamePattern, this configuration specifies the tables to be skipped.
 By default the black list is empty which means no tables will be skipped.
 
+**numSplit**: Number of splits per table. By default is 1
```
Comment (albertshau) [inline docs/MultiTableDatabase-batchsource.md]: docs should use the property names that appear in the UI. Can you update this doc to do that?

File: docs/MultiTableDatabase-batchsource.md [hunk 2]
```diff
@@ -62,45 +63,59 @@ This example reads from all tables in the 'customers' database on host 'host123.
...
+    +-----+--------+----------+--------+
+    | 0   | 0      | shirt123 | view   |
+    | 1   | 0      | carxyz   | view   |
+    | 2   | 0      | shirt123 | buy    |
+    | 3   | 0      | coffee   | view   |
+    | 4   | 1      | cola     | buy    |
+    | 5   | 1      | pepsi    | buy    |
+    +-----+--------+----------+--------+
+
+You will have 4 splits (2 per each table) with such queries:
```
Comment (albertshau) [inline docs/MultiTableDatabase-batchsource.md:96]: need to mention that the ID column is the primary key for both tables.

File: src/main/java/io/cdap/plugin/format/DBTableInfo.java [hunk 3]
```diff
@@ -18,23 +18,33 @@
...
-    return name;
+  public DBTableName getDbTableName() {
+    return dbTableName;
   }
 
   public Schema getSchema() {
     return schema;
   }
+
+  public List<String> getPrimaryColumns() {
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/format/DBTableInfo.java]: nit: primaryColumns -> primaryKey

File: src/main/java/io/cdap/plugin/format/DBTableName.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package io.cdap.plugin.format;
+
+/**
+ * Information about a DB table name.
+ */
+public class DBTableName {
+  private String db;
+  private String tableName;
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/format/DBTableName.java]: nit: should be consistent and use 'dbName' and 'tableName', or 'db' and 'table'. I think it's better without the 'Name'

File: src/main/java/io/cdap/plugin/format/DBTableRecordReader.java [hunk 5]
```diff
@@ -111,7 +111,7 @@ public StructuredRecord getCurrentValue() throws IOException {
   }
 
   @Override
-  public float getProgress() {
+  public float getProgress() throws IOException {
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/format/DBTableRecordReader.java:114]: why was this added?

File: src/main/java/io/cdap/plugin/format/DBTableRecordReader.java [hunk 6]
```diff
@@ -157,4 +157,15 @@ public void close() throws IOException {
       throw new IOException(exception);
     }
   }
+
+  private String getQuery() {
+    if (split.getLowerClause() != null && split.getUpperClause() != null) {
+      // Now append the conditions associated with our split.
+      // Build the WHERE clauses associated with the data split.
+      String conditionClauses = "WHERE (( " + split.getLowerClause() + (" ) AND ( ") + split.getUpperClause() + " ))";
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/format/DBTableRecordReader.java]: This type of logic seems like it should be encapsulated in the split class and exposed in a getWhereClause() method.

File: src/main/java/io/cdap/plugin/format/MultiTableConf.java [hunk 7]
```diff
@@ -175,11 +187,22 @@ public Connection getConnection() throws SQLException {
...
-    StringBuilder query = new StringBuilder(selectClause).append(table);
+    String query = selectClause + table;
+    if (whereClause != null && !whereClause.isEmpty()) {
+      query += " " + whereClause;
+    }
+
+    return query;
+  }
+
+  public String appendWhereClauseWithSplit(String selectClause, String table, String split) {
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/format/MultiTableConf.java]: this doesn't belong in the config. The query is logic internal to the input format.

File: src/main/java/io/cdap/plugin/format/MultiTableConf.java [hunk 8]
```diff
@@ -146,7 +151,14 @@ public String getTableNameField() {
   }
 
   @Nullable
-  public String getWhereClause() { return whereClause; }
+  public Integer getNumSplits() {
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/format/MultiTableConf.java]: a better name is splitsPerTable, this sounds like the total number of splits

File: src/main/java/io/cdap/plugin/format/MultiTableDBConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+/**
+ * Allows to specify and access connection configuration properties of {@link Configuration}.
+ */
+public class MultiTableDBConfiguration extends DBConfiguration {
+  private static final String PLUGIN_CONF_FIELD = "multi.db.plugin.conf";
+  private static final String DRIVER_FIELD = "multi.db.jdbc.connection";
+  private static final String INFO_FIELD = "multi.db.jdbc.dbinfo";
+
+  private static final Type PLUGIN_CONF_TYPE = new TypeToken<MultiTableConf>() {
+  }.getType();
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/format/MultiTableDBConfiguration.java]: nit: can be on previous line: `{ }.getType()`

File: src/main/java/io/cdap/plugin/format/MultiTableDBInputFormat.java [hunk 10]
```diff
@@ -131,15 +170,63 @@
...
+           statement.executeQuery(getBoundingValsQuery(info.getDbTableName().fullTableName(), columnName))) {
+      results.next();
+
+      // Based on the type of the results, use a different mechanism
+      // for interpolating split points (i.e., numeric splits, text splits,
+      // dates, etc.)
+      int sqlDataType = results.getMetaData().getColumnType(1);
+      DBSplitter splitter = getSplitter(sqlDataType);
+      if (null == splitter) {
+        LOG.warn("Failed to create internal splits for table " + info.getDbTableName().fullTableName() +
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/format/MultiTableDBInputFormat.java]: warn -> info

File: widgets/MultiTableDatabase-batchsource.json [hunk 11]
```diff
@@ -66,6 +66,14 @@
           "widget-type": "textbox",
           "label": "Enable Auto Commit",
           "name": "enableAutoCommit"
+        },
+        {
+          "widget-type": "textbox",
+          "label": "Number of splits per table",
```
Comment (albertshau) [inline widgets/MultiTableDatabase-batchsource.json]: Splits Per Table

File: src/main/java/io/cdap/plugin/format/MultiTableDBInputFormat.java [hunk 12]
```diff
@@ -42,18 +40,16 @@
...
 import java.util.Collection;
+import java.util.Collections;
 import java.util.List;
+import java.util.stream.Collectors;
 
 /**
  * Input format that reads from multiple tables in a database using JDBC. Similar to Hadoop's DBInputFormat.
  */
-public class  MultiTableDBInputFormat extends InputFormat<NullWritable, StructuredRecord> {
+public class MultiTableDBInputFormat<T extends DBWritable> extends DataDrivenDBInputFormat<T> {
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/format/MultiTableDBInputFormat.java]: I don't understand how this is working anymore if the input format is returning a DBWritable instead of a StructuredRecord. The source still expects the inputformat to return StructuredRecord

File: docs/MultiTableDatabase-batchsource.md (hunk 1 above)
Comment (mlozbin-cybervisiontech) [reply docs/MultiTableDatabase-batchsource.md]: done

File: docs/MultiTableDatabase-batchsource.md (hunk 2 above)
Comment (mlozbin-cybervisiontech) [reply docs/MultiTableDatabase-batchsource.md:96]: done

File: src/main/java/io/cdap/plugin/format/DBTableInfo.java (hunk 3 above)
Comment (mlozbin-cybervisiontech) [reply src/main/java/io/cdap/plugin/format/DBTableInfo.java]: done

File: src/main/java/io/cdap/plugin/format/DBTableName.java (hunk 4 above)
Comment (mlozbin-cybervisiontech) [reply src/main/java/io/cdap/plugin/format/DBTableName.java]: done

File: src/main/java/io/cdap/plugin/format/DBTableRecordReader.java (hunk 6 above)
Comment (mlozbin-cybervisiontech) [reply src/main/java/io/cdap/plugin/format/DBTableRecordReader.java]: done

File: src/main/java/io/cdap/plugin/format/MultiTableConf.java (hunk 8 above)
Comment (mlozbin-cybervisiontech) [reply src/main/java/io/cdap/plugin/format/MultiTableConf.java]: done

File: src/main/java/io/cdap/plugin/format/MultiTableDBConfiguration.java (hunk 9 above)
Comment (mlozbin-cybervisiontech) [reply src/main/java/io/cdap/plugin/format/MultiTableDBConfiguration.java]: done

File: src/main/java/io/cdap/plugin/format/MultiTableDBInputFormat.java (hunk 10 above)
Comment (mlozbin-cybervisiontech) [reply src/main/java/io/cdap/plugin/format/MultiTableDBInputFormat.java]: done

File: widgets/MultiTableDatabase-batchsource.json (hunk 11 above)
Comment (mlozbin-cybervisiontech) [reply widgets/MultiTableDatabase-batchsource.json]: done

File: src/main/java/io/cdap/plugin/format/DBTableRecordReader.java (hunk 5 above)
Comment (mlozbin-cybervisiontech) [reply src/main/java/io/cdap/plugin/format/DBTableRecordReader.java:114]: getLength() method which uses in this method can throw IOException,

File: src/main/java/io/cdap/plugin/format/MultiTableConf.java (hunk 7 above)
Comment (mlozbin-cybervisiontech) [reply src/main/java/io/cdap/plugin/format/MultiTableConf.java]: Moved  appendWhereClauseWithSplit() logic to DBTableRecordReader and appendWhereClause() logic to MultiTableDBInputFormat class.

File: src/main/java/io/cdap/plugin/format/MultiTableDBInputFormat.java (hunk 12 above)
Comment (mlozbin-cybervisiontech) [reply src/main/java/io/cdap/plugin/format/MultiTableDBInputFormat.java]: Reverted to extend of InputFormat, and copied getSplitter() method from DataDrivenDBInputFormat.

**Comment by mlozbin-cybervisiontech** [comment]
@albertshau I addressed all your comments, could you, please, take a look.

File: docs/MultiTableDatabase-batchsource.md [hunk 13]
```diff
@@ -51,7 +51,7 @@ By default reads all tables matching the tableNamePattern.
 **blackList**: Used in conjunction with tableNamePattern, this configuration specifies the tables to be skipped.
 By default the black list is empty which means no tables will be skipped.
 
-**numSplit**: Number of splits per table. By default is 1
+**splitsPerTable**: The number of splits per table. By default is 1
```
Comment (albertshau) [inline docs/MultiTableDatabase-batchsource.md]: as mentioned before, please change these properties to use the UI label. 'Splits Per Table' 'Where Clause', etc.

File: src/main/java/io/cdap/plugin/format/DBTableRecordReader.java [hunk 14]
```diff
@@ -159,13 +159,15 @@ public void close() throws IOException {
...
-    if (split.getLowerClause() != null && split.getUpperClause() != null) {
-      // Now append the conditions associated with our split.
-      // Build the WHERE clauses associated with the data split.
-      String conditionClauses = "WHERE (( " + split.getLowerClause() + (" ) AND ( ") + split.getUpperClause() + " ))";
-      return dbConf.appendWhereClauseWithSplit("SELECT * FROM ", tableName.fullTableName(), conditionClauses);
+    String query = "SELECT * FROM " + tableName.fullTableName() + " ";
+    String whereClause = dbConf.getWhereClause();
+
+    if (whereClause != null && !whereClause.isEmpty()) {
+      query += split.getWhereClause() + " AND (" + whereClause.substring("where".length()) + ")";
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/format/DBTableRecordReader.java]: it's a little cleaner to make split.getWhereClause() not actually return 'where', so that you don't have to do a substring. Then this becomes:

```
if (...) {
  query += whereClause + " AND " + split.getWhereClause();
} else {
  query += "WHERE " + split.getWhereClause();
}

**Review by albertshau — COMMENTED** [review]
minor comments, please fix and squash.

File: docs/MultiTableDatabase-batchsource.md (hunk 13 above)
Comment (mlozbin-cybervisiontech) [reply docs/MultiTableDatabase-batchsource.md]: done

File: src/main/java/io/cdap/plugin/format/DBTableRecordReader.java (hunk 14 above)
Comment (mlozbin-cybervisiontech) [reply src/main/java/io/cdap/plugin/format/DBTableRecordReader.java]: done

**Comment by mlozbin-cybervisiontech** [comment]
@albertshau I addressed all your comments and squashed commits, could you, please, merge.

**Review by albertshau — APPROVED** [review]
